### PR TITLE
feat(GUI): link to drivers when clicking a driverless usbboot device

### PR DESCRIPTION
**Pull request by jviotti**
Step 2 until we support installing the drivers from within Etcher. This
also introduces an "Open drive link" Mixpanel event.

See: https://www.raspberrypi.org/documentation/hardware/computemodule/cm-emmc-flashing.md
See: https://github.com/resin-io/etcher/pull/1892
Change-Type: patch
Changelog-Entry: Download usbboot drivers installer when clicking a driverless usbboot device on Windows.
Signed-off-by: Juan Cruz Viotti <jv@jviotti.com>

### CODE REVIEW

**Comment by jviotti** [comment]
If you want to test this, you can mock a driverless usbboot device on any platform with the following patch:

```diff
diff --git a/lib/shared/sdk/usbboot/usb.js b/lib/shared/sdk/usbboot/usb.js
index d0e2d98f..56849527 100644
--- a/lib/shared/sdk/usbboot/usb.js
+++ b/lib/shared/sdk/usbboot/usb.js
@@ -72,6 +72,14 @@ exports.listDevices = () => {
     return device
   })

+  devices.push({
+    accessible: false,
+    deviceDescriptor: {
+      idVendor: 0x0a5c,
+      idProduct: 0x2764
+    }
+  })
+
   // Include driverless devices into the list of USB devices.
   if (process.platform === 'win32') {
     const winusbDriverGenerator = require('winusb-driver-generator')
```

**Comment by lurch** [comment]
Could you include a screenshot of what this looks like please?

File: lib/gui/components/drive-selector/controllers/drive-selector.js [hunk 1]
```diff
@@ -27,7 +27,8 @@ const selectionState = require('../../../../shared/models/selection-state')
 module.exports = function (
   $q,
   $uibModalInstance,
-  WarningModalService
+  WarningModalService,
```
Comment (Shou) [inline lib/gui/components/drive-selector/controllers/drive-selector.js]: Should we remove this? Or are there plans for a warning modal?

Comment (jviotti) [reply lib/gui/components/drive-selector/controllers/drive-selector.js]: There are other functions in this controller that use this service.

**Comment by jviotti** [comment]
> Could you include a screenshot of what this looks like please?

Nothing really to screenshot. When you click on the device that said "Missing drivers", then your browser opens up.

**Comment by jviotti** [comment]
> Looks good otherwise, but maybe we should use a modal with the link instead of downloading it on click? Might surprise the user e.g. they can think it's a virus.

Hm, maybe link to a landing page that links to the drivers instead?

**Comment by lurch** [comment]
Perhaps we could underline the link and display it in blue, otherwise I guess there's no CTA that it's something that can be clicked on?

**Comment by lurch** [comment]
> Hm, maybe link to a landing page that links to the drivers instead?

Or I guess we could display a dialog which says something like "We've detected that you've plugged in a Compute Module but you don't have the necessary driver installed. Click yes below to download the driver". Just so that the user gets a chance to know what's going on before their browser starts downloading an unknown executable file (while also giving them the chance to say No if they _don't_ want to download the drivers at this point).

**Comment by jviotti** [comment]
Alright. I updated the PR so the "Missing drivers" entry is not disabled (so the user knows he can click it), plus added a modal before opening the link in the user's browser:

screen shot 2018-01-04 at 2 55 16 pm
screen shot 2018-01-04 at 2 55 19 pm

**Comment by alexandrosm** [comment]
Hey @jviotti -- this UX is still really scary, and users don't really know
why to click. Can you bring this to a product call pls?

-- 

*Alexandros Marinos*

Founder & CEO, Resin.io


+1 206-637-5498

@alexandrosm

**Comment by konmouz** [comment]
@jviotti, I was thinking if it makes sense to have a download icon or even icon+text CTA so the user can download the relevant drivers. Also red feels a bit too much, we can have a smaller + message in orange.

**Comment by jhermsmeier** [comment]
Rebased, and updated according to @konmouz' feedback;

- Added friendlier modal, leading to the driver download page
- Added CTA link next to "Missing Drivers" note on CM in drive list

screen shot 2018-05-02 at 23 24 31

screen shot 2018-05-02 at 23 24 35

File: lib/gui/app/components/drive-selector/controllers/drive-selector.js [hunk 2]
```diff
@@ -98,9 +101,51 @@ module.exports = function (
...
+   *
+   * @example
+   * DriveSelectorController.installMissingDrivers({
+   *   linkTitle: 'Go to example.com',
+   *   linkMessage: 'Examples are great, right?',
+   *   linkCTA: 'Call To Action',
+   *   link: 'https://example.com'
+   * });
+   */
+  this.installMissingDrivers = (drive) => {
```
Comment (lurch) [inline lib/gui/app/components/drive-selector/controllers/drive-selector.js]: Is `drive` really a suitable name for this parameter??

File: lib/gui/app/components/drive-selector/controllers/drive-selector.js [hunk 3]
```diff
@@ -98,9 +101,51 @@ module.exports = function (
...
+  /**
+   * @summary Prompt the user to install missing usbboot drivers
+   * @function
+   * @public
+   *
+   * @param {Object} drive - drive
+   * @returns {Promise} - resolved promise
+   *
+   * @example
+   * DriveSelectorController.installMissingDrivers({
```
Comment (lurch) [inline lib/gui/app/components/drive-selector/controllers/drive-selector.js]: Perhaps `downloadMissingDrivers` is a better name?

File: lib/gui/app/components/drive-selector/controllers/drive-selector.js [hunk 4]
```diff
@@ -98,9 +101,51 @@ module.exports = function (
...
+   *   linkTitle: 'Go to example.com',
+   *   linkMessage: 'Examples are great, right?',
+   *   linkCTA: 'Call To Action',
```
Comment (lurch) [inline lib/gui/app/components/drive-selector/controllers/drive-selector.js]: Strictly speaking, `linkCTA` isn't used in this function.

**Comment by lurch** [comment]
Looks like a nice UX, but presumably this "Install" link is only shown in the drive-selector when Etcher is running on Windows? (given that the dialog mentions "Window installer")

File: lib/gui/app/components/drive-selector/templates/drive-selector-modal.tpl.html [hunk 5]
```diff
@@ -23,7 +23,8 @@ <h4 class="list-group-item-heading">{{ drive.description }}
             <span class="word-keep"
               ng-show="drive.size"> - {{ drive.size | closestUnit }}</span>
           </h4>
-          <p class="list-group-item-text">{{ drive.displayName }}</p>
+          <p class="list-group-item-text" ng-if="!drive.link">{{ drive.displayName }}</p>
+          <p class="list-group-item-text" ng-if="drive.link">{{ drive.displayName }} - <b><a ng-click="modal.installMissingDrivers(drive)">{{ drive.linkCTA }}</a></b></p>
```
[the remark this answers is not in the document]
Comment (Shou) [reply lib/gui/app/components/drive-selector/templates/drive-selector-modal.tpl.html]: It won't be shown on Linux so we should be fine, but ideally it'd be hidden if that env var is set

**Pull request merged**